### DISCUSSION TITLE
Override autonaming with @cell decorator

### DISCUSSION THREAD

**Discussion by FJohlinger** [gdsfactory/gdsfactory#1547] [Q&A] [answered]
Hi, I've been writing some functions that import components from gds and yaml files into a component. I'm using the `@gf.cell` decorator to add the component to the cache. However, I would like to name the component myself and not use the 'unique naming' functionality. I have a system that can create unique component names, so I am not worried about duplicates. 

I cannot use the automatically created names, since I want to have multiple modules, which each having the same function names within it. E.g. a module per wavelength and then within each module a function creating a 1x2 mmi. This way I can easily switch between wavelengths, while keeping mostly the same code for a similar layout. However, since the name of the component is based on the function wrapping around the gds import (the gdsname is hardcoded in for ease of use, so no function parameters), I get non-unique names.

I saw that there was a `autoname = False` option added as a parameter to `gf.cell`, which doesn't seem usable with the decorator version, though. So, I would need to wrap all my functions in the cell function manually, which just doesn't make for very pretty code. Is there any cleaner solution to stop the `@gf.cell` decorator renaming my components?

**Comment by joamatab** [comment]
Hi @FJohlinger,


to avoid the autonaming from `@cell`  we have 2 options:


1. Use `autoname=False`

```python

c = gf.components.mzi(autoname=False)

```

2. Add `component.imported_gds = True` so autoname won't happen. This is what we do in import_gds

**Reply by FJohlinger** [reply]
The first option is one I'd like to avoid and I can't seem to make the second option to work, when I write it as below. 

```
import gdsfactory as gf

@gf.cell
def mmi1x2() -> gf.Component:
	comp = gf.import_gds('mm1x2.gds', read_metadata=True)
	comp.name = 'my_own_unique_name'
	comp.imported_gds = True
	return comp
```

Then when I use 
```
comp = mmi1x2()
print(comp.name)
```
I still get `mmi1x2` as output instead of 'my_own_unique_name'. What am I doing wrong?

**Comment by joamatab** [comment]
How about using?

```python

comp = mmi1x2(autoname=False)
print(comp.name)

```

**Reply by FJohlinger** [reply]
I was trying to avoid having to add parameters to the function call, so though I think that works, I'll be using the solution I found below.

**Comment by FJohlinger** [comment, marked as answer]
Got it to work when creating a new component and adding the imported component to it:

```
import gdsfactory as gf

@gf.cell
def mmi1x2() -> gf.Component:
	comp = gf.Component()
	comp << gf.import_gds('mm1x2.gds', read_metadata=True)
	comp.name = 'my_own_unique_name'
	comp.imported_gds = True
	return comp
```

**Comment by FJohlinger** [comment]
It seems like the solution above is no longer working with the changes to the @gf.cell decorator over the last few months. I've tried the following approaches

```
import gdsfactory as gf

@gf.cell(autoname = False)
def mmi1x2() -> gf.Component:
	comp = gf.import_gds('my_own_unique_name.gds', read_metadata=True)
	return comp
```

=> This will cause the component to receive the name 'Unnamed_<some_hash>'

```
import gdsfactory as gf

@gf.cell(autoname = False)
def mmi1x2() -> gf.Component:
	comp = gf.Component()
	comp << gf.import_gds('my_own_unique_name.gds', read_metadata=True)
	comp.rename(name='my_own_unique_name')
	return comp
```

=> When this function is called twice, the first component is created as wanted (just one cell, called "my_own_unique_name" ), the second one is called "my_own_unique_name$1" and has a subcell called "my_own_unique_name". Also, when saving the layout that these get added to to gds I get a "Duplicated cell names" warning.

```
import gdsfactory as gf

@gf.cell(autoname = False)
def mmi1x2() -> gf.Component:
	comp = gf.Component()
	comp << gf.import_gds('my_own_unique_name.gds', read_metadata=True)
	comp.rename(name='my_own_unique_name')
        comp = comp.flatten()
	return comp
```

=> When this function is called twice, the first component is created as wanted (just one cell, called "my_own_unique_name" ), the second one is called "my_own_unique_name$1" (no subcell any more, as expected). No "Duplicated cell names" warning for this version.

```
import gdsfactory as gf

@gf.cell(autoname = False)
def mmi1x2() -> gf.Component:
	comp = gf.Component()
	comp << gf.import_gds('my_own_unique_name.gds', read_metadata=True)
	comp.rename(name='my_own_unique_name', cache = False)
        comp = comp.flatten()
	return comp
```

=> When this function is called twice, both components are created as wanted (just one cell, called "my_own_unique_name"  referred to twice on the layout). However, when saving the layout that these get added to to gds I get a "Duplicated cell names" warning.

Is there a better way to approach this situation? I would like to avoid the $1, $2, ... postfix, because that would bloat the size of the gds over time, however the duplicate cell warning is something I'd like to avoid as well, because that might obscure real duplicate cell problems elsewhere in the design.

**Comment by joamatab** [comment]
It seems like you are writing your own custom `import_gds`

the issue you are having is because a check we added on `cell` decorator that makes a copy if the component is already on the cache

I disabled that so that you can use your custom `import_gds`  [here](https://github.com/gdsfactory/gdsfactory/pull/2661)

also, why not just run like this?


```
def mmi1x2() -> gf.Component:
	return gf.import_gds('my_own_unique_name.gds', read_metadata=True)
````

@tvt173

**Reply by FJohlinger** [reply]
Thanks for replying so quickly. Either version you suggested creates the right cells:

```
def mmi1x2() -> gf.Component:
	return gf.import_gds('my_own_unique_name.gds', read_metadata=True)
```

or 

```
@gf.cell(autoname=False, copy_if_cached=False)
def mmi1x2() -> gf.Component:
    return gf.import_gds('my_own_unique_name.gds', read_metadata=True)
```

However, both versions also still give me the 'Duplicated cell names' warning when I have two copies added to my layout and I try to write to gds. If you think that should be the expected behaviour, I'll stick with one of these for now.

**Reply by joamatab** [reply]
interesting, yes, i was able to reproduce your issue, thanks for sharing it

Here is a workaround to avoid duplicated cells 

```python
from functools import cache

@cache
def mmi1x2() -> gf.Component:
	return gf.import_gds('my_own_unique_name.gds', read_metadata=True)
```

**Reply by FJohlinger** [reply]
That works great. Thanks!

**Reply by joamatab** [reply]
Awesome!
with the new gdsfactory release the duplicated cells should have been fixed too

**Reply by FJohlinger** [reply]
I think the `@cache` solution works, as well as using no decorator.

When I update to the newest version (v7.23.0) of gdsfactory and use `@gf.cell(autoname=False, copy_if_cached=False)` it seems like functions with the same name in different modules (e.g. `module1.some_function` returning 'comp1' and `module2.some_function` returning 'comp2') return the component imported from the function which gets called first. When the function in the second module is called afterwards it doesn't return the component that it is supposed to import, but instead seems to refer to the function in the first module and returns that component instead. So if we first use `module1.some_function`, calling `module2.some_function` afterwards returns comp1 instead of comp2.

**Reply by joamatab** [reply]
yes, how about we include the module function as part of the hashing function for the name?

**Reply by joamatab** [reply]
why not do this?

in module1
```

@gf.cell(include_module=True)
def straight(**kwargs):
  return gf.components.straight(**kwargs)
````

**Reply by FJohlinger** [reply]
Thanks a lot, that's doing everything now! I'm using the decorator with the following options:
```
@gf.cell(
    autoname=False, 
    copy_if_cached=False, 
    include_module=True
)
```

**Reply by joamatab** [reply]
awesome! thank you for your suggestion of adding `include_module` in the cell decorator some time ago
